### PR TITLE
Кудряшова Ирина. Задача 1. Вариант 9. Скалярное произведение векторов.

### PR DESCRIPTION
**Pull request by Irina2004-tech**
Описание последовательной задачи
Задача заключается в вычислении скалярного произведения двух векторов. Алгоритм поэлементно перемножает соответствующие элементы векторов, а затем суммирует полученные результаты.

Описание MPI задачи
Параллельная задача решается на нескольких процессах MPI. На момент начала рассмотрения задачи мы имеем двухмерный вектор чисел, который содержит два входных вектора. Эти векторы делятся на сегменты в зависимости от количества процессов. Затем каждый сегмент передается соответствующему процессу для дальнейшей обработки. Каждый процесс получает свой сегмент данных, выполняет поэлементное умножение векторов и суммирует результаты. После этого результаты со всех процессов переходят в корневой процесс, где суммируются для получения результата, который и будет ответом.

### CODE REVIEW

File: tasks/seq/kudryashova_i_vector_dot_product/src/vectorDotProductSeq.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#include "seq/kudryashova_i_vector_dot_product/include/vectorDotProductSeq.hpp"
+
+#include <random>
+
+static int seedOffset = 0;
+std::vector<int> kudryashova_i_vector_dot_product::GetRandomVector(int size) {
```
Comment (allnes) [inline tasks/seq/kudryashova_i_vector_dot_product/src/vectorDotProductSeq.cpp:6]: please move this fucntion to test files

File: tasks/mpi/kudryashova_i_vector_dot_product/src/vectorDotProductMPI.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
+#include "mpi/kudryashova_i_vector_dot_product/include/vectorDotProductMPI.hpp"
+
+#include <boost/mpi.hpp>
+#include <random>
+
+static int seedOffset = 0;
+
+std::vector<int> kudryashova_i_vector_dot_product_mpi::getRandomVector(int size) {
```
Comment (allnes) [inline tasks/mpi/kudryashova_i_vector_dot_product/src/vectorDotProductMPI.cpp:8]: please move this function to test files

**Review by allnes — CHANGES_REQUESTED** [review]

File: tasks/mpi/kudryashova_i_vector_dot_product/func_tests/vectorDotProductMPIFuncTests.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      taskDataPar->inputs.emplace_back(reinterpret_cast<uint8_t*>(global_vector[i].data()));
+    }
+    taskDataPar->inputs_count.emplace_back(global_vector[0].size());
+    taskDataPar->inputs_count.emplace_back(global_vector[1].size());
+    taskDataPar->outputs.emplace_back(reinterpret_cast<uint8_t*>(result.data()));
+    taskDataPar->outputs_count.emplace_back(result.size());
+  }
+  kudryashova_i_vector_dot_product_mpi::TestMPITaskParallel testMpiTaskParallel(taskDataPar);
+  ASSERT_EQ(testMpiTaskParallel.validation(), true);
+}
```
Comment (aobolensk) [inline tasks/mpi/kudryashova_i_vector_dot_product/func_tests/vectorDotProductMPIFuncTests.cpp:127]: Please add test with empty vectors and vectors of the length 1

**Review by ArinaTs — COMMENTED** [review]
add a test for unequal vectors

File: tasks/mpi/kudryashova_i_vector_dot_product/func_tests/vectorDotProductMPIFuncTests.cpp (hunk 3 above)
Comment (Irina2004-tech) [reply tasks/mpi/kudryashova_i_vector_dot_product/func_tests/vectorDotProductMPIFuncTests.cpp:239]: done

File: tasks/mpi/kudryashova_i_vector_dot_product/src/vectorDotProductMPI.cpp (hunk 2 above)
Comment (Irina2004-tech) [reply tasks/mpi/kudryashova_i_vector_dot_product/src/vectorDotProductMPI.cpp]: done

File: tasks/seq/kudryashova_i_vector_dot_product/src/vectorDotProductSeq.cpp (hunk 1 above)
Comment (Irina2004-tech) [reply tasks/seq/kudryashova_i_vector_dot_product/src/vectorDotProductSeq.cpp]: done

**Review by ArinaTs — APPROVED** [review]
good job

File: tasks/mpi/kudryashova_i_vector_dot_product/func_tests/vectorDotProductMPIFuncTests.cpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+std::vector<int> GetRandomVector(int size) {
+  std::vector<int> vector(size);
+  std::srand(static_cast<unsigned>(time(nullptr)) + ++seedOffset);
+  for (int i = 0; i < size; ++i) {
+    vector[i] = std::rand() % 100 + 1;
+  }
+  return vector;
+}
+
```
Comment (ascannel) [inline tasks/mpi/kudryashova_i_vector_dot_product/func_tests/vectorDotProductMPIFuncTests.cpp:19]: add tests that asserts false when trying to dot product empty vectors

File: tasks/mpi/kudryashova_i_vector_dot_product/src/vectorDotProductMPI.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+bool kudryashova_i_vector_dot_product_mpi::TestMPITaskSequential::pre_processing() {
+  internal_order_test();
+
+  input_.resize(taskData->inputs.size());
+  for (unsigned long i = 0; i < input_.size(); ++i) {
+    auto* tempPtr = reinterpret_cast<int*>(taskData->inputs[i]);
+    input_[i] = std::vector<int>(taskData->inputs_count[i]);
+    std::copy(tempPtr, tempPtr + taskData->inputs_count[i], input_[i].begin());
+  }
+  result = 0;
```
Comment (ascannel) [inline tasks/mpi/kudryashova_i_vector_dot_product/src/vectorDotProductMPI.cpp:22]: remove the initialization of the resulting value to zero; it is set when you specify result{}

File: tasks/mpi/kudryashova_i_vector_dot_product/include/vectorDotProductMPI.hpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+#include <boost/mpi/collectives.hpp>
+#include <boost/mpi/communicator.hpp>
+#include <memory>
+#include <numeric>
+#include <string>
```
Comment (ascannel) [inline tasks/mpi/kudryashova_i_vector_dot_product/include/vectorDotProductMPI.hpp:6]: remove unnecessary includes

**Review by ascannel — CHANGES_REQUESTED** [review]

**Review by allnes — APPROVED** [review]

**Review by ascannel — APPROVED** [review]

File: tasks/mpi/kudryashova_i_vector_dot_product/func_tests/vectorDotProductMPIFuncTests.cpp (hunk 3 above)
Comment (Irina2004-tech) [reply tasks/mpi/kudryashova_i_vector_dot_product/func_tests/vectorDotProductMPIFuncTests.cpp:280]: done

**Review by aobolensk — APPROVED** [review]
